### DISCUSSION TITLE
How to do multithreaded decompilation in python

### DISCUSSION THREAD

**Discussion by dzik87** [NationalSecurityAgency/ghidra#2903] [Q&A] [answered]
Hi, 

I am wondering if there is any way to make decompilation in python using all available cores.
This would greatly speedup entire process.

Part of code I am using now:
```
ifc = DecompInterface()
    ifc.setOptions(options)
    ifc.openProgram(currentProgram)
    for func in currentProgram.functionManager.getFunctions(1): 
        monitor.incrementProgress(1)
        monitor.setShowProgressValue(True)
            
        monitor.setMessage("Analyzing 0x{}".format(func.getEntryPoint()))
        res = ifc.decompileFunction(func, 60, monitor)
        high_func = res.getHighFunction()
        if high_func:
            lsm = high_func.getLocalSymbolMap()
            
            hfdb = HighFunctionDBUtil()
            hfdb.commitParamsToDatabase(high_func, True, SourceType.USER_DEFINED)
            hfdb.commitReturnToDatabase(high_func, SourceType.USER_DEFINED)
            hfdb.commitLocalNamesToDatabase(high_func, SourceType.USER_DEFINED)
```

To decompile entire app it required around 30minutes and i would like to speedup this process.

**Comment by astrelsky** [comment, marked as answer]
There is a parallel decompiler class you can use. It is used in some of the analyzers that use the decompiler. I forget the name of the class. It has parallel in the name though.

Edit: see [ParallelDecompiler](https://ghidra.re/ghidra_docs/api/ghidra/app/decompiler/parallel/ParallelDecompiler.html)

Also I haven't checked your example to be sure but be aware of the `Disposable` interface. Failure to call `dispose` may be a resource leak.

**Reply by dzik87** [reply]
Thank you. 
This should be enough for my use

**Comment by dragonmacher** [comment]
Some uses of the `ParallelDecompiler` can be find in many analyzers and scripts, such as the `CompareFunctionSizesScript` and the `DecompilerSwitchAnalyzer`.

**Reply by dzik87** [reply]
Thanks.
Examples were really helpfull
